pandas/tests/tools/test_merge.py

```diff
@@ -1237,8 +1233,177 @@ def f():
                             .09604978, -.06524096, .03532373,
                             .03025441, .036997, None, None]
             ))
-            .set_index(["household_id", "asset_id", "t"]))
+            .set_index(["household_id", "asset_id", "t"])
+            .reindex(columns=['share', 'log_return']))
+
+        result = (merge(household.reset_index(), log_return.reset_index(),
+                on=['asset_id'], how='outer')
+          .set_index(['household_id', 'asset_id', 't']))
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
+    
+        # Self join
+        matrix = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2, 2, 3],
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM','PM','IP','AM','OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444]),
+                columns=['Origin', 'Destination', 'Period',
+                         'TripPurp', 'Trips'])
+            .set_index(['Origin', 'Destination', 'Period', 'TripPurp']))
+        
+        distances = (
+            pd.DataFrame(
+                dict(Origin=     [1, 1, 2, 2, 3, 3, 5],
+                     Destination=[1, 2, 1, 2, 1, 2, 6],
+                     Period=['AM','PM','IP','AM','OP','IP', 'AM'],
+                     LinkType=['a', 'a', 'c', 'b', 'a', 'b', 'a'],
+                     Distance=[100, 80, 90, 80, 75, 35, 55]),
+                columns=['Origin', 'Destination', 'Period', 
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination','Period', 'LinkType']))
+        
+        expected = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2, 2, 3],
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM','PM','IP','AM','OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444],
+                     Trips_joined=[1987, 3647, 2470, 4296, 4444]),
+                columns=['Origin', 'Destination', 'Period',
+                         'TripPurp', 'Trips', 'Trips_joined'])
+            .set_index(['Origin', 'Destination', 'Period', 'TripPurp']))
+        
+        result = matrix.join(matrix, how='inner', rsuffix='_joined') 
+        assert_frame_equal(result, expected)
+        
+        #Left join
+        expected = (
+            pd.DataFrame(
+                dict(Origin=     [1, 1, 2, 2, 3],
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM','PM','IP', 'AM', 'OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444],
+                     Distance=[100, 80, 90, np.nan, 75]),
+                columns=['Origin', 'Destination', 'Period', 'TripPurp', 
```

Comment:
Nevertheless, the current commit includes all the requested changes. You can review them whenever possible